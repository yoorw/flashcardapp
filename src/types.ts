// defines the flashcard objects that the app stores and displays
export interface Card {
  answer: string,   // answer to the question
  question: string, // question prompt
  subject: string   // subject of the question and answer
}

// the shape of the state that CardContext returns
export interface CardState {
  // the array of Card objects
  cards: Card[],

  // the index of the currently displayed card object
  current: number,

  // the dispatch function that accepts actions
  // actions are handled by the reducer in CardContext
  dispatch: (action: any) => void
}

// the types of action that the reducer in CardContext will handle
export enum CardActionTypes {
  next = 'next',
};

export type CardAction = 
  // moves to the next card
  | { type: CardActionTypes.next }

  // The Stats for a single question
  export interface Stats {
    // number of times user has gotten it right
    right: number,

    // number of times user has gotten it wrong
    wrong: number,

    // number of times user has seen thew question but skipped it
    // instead of answering it
    skip: number
  };

  // interface with a string index signature
  // each string is expected to return an object that fits the Stats interface
  // the string that we will use for a signature is the question from a Card object
  export interface StatsType {
    [key: string]: Stats
  };

  // The StatsDispatch function 
  interface StatsDisaptch {
    dispatch: (action: StatsAction) => void
  };

  // union type. The stats state will have a Stats object for any given key
  // except dispatch will return the StatsDispatch function
  export type StatsState = StatsType & StatsDisaptch;

  // an enum listing the three types of StatsAction
  // A User can get a question right, wrong, or skip it
  export enum StatsActionType {
    right = 'right',
    skip = 'skip',
    wrong = 'wrong'
  };

  // Stats Action - takes a question from a card
  export type StatsAction = {
    type: StatsActionType,
    question: string
  };
