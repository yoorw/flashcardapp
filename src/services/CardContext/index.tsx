import React, {createContext, useEffect, useReducer} from "react";
import { getAdditionCard } from "../../shared/utils";
import {Card, CardState} from '../../types';
import {saveCards} from '../Save';
import {getInitialState, getNext} from './services';


// the reducer handles actions 
export const reducer = (state: CardState, action: any) => {
  // switch statement looks at the action type
  // if there is a case that matches the type it will run that code 
  // otherwise it will run the default case
  switch(action.type) {
    // action delete
    case 'delete': {
      let {cards, current} = state;
      // the question is the unique identifier of a card
      const {question} = action;

      // creating a new array of cards by spreading the current array of cards
      const newCards = [...cards];

      // finds the index of the target card
      const index = newCards.findIndex(card => card.question === question);

      // splice removes the target card from the array
      newCards.splice(index, 1);

      // current tells the components what card to display
      // decrement current
      current = current -1;

      // don't pass -1 as current
      if(current < 0) current = 0;

      // spread the old state
      // add the new value of current
      // and return the newCards array as the value of cards
      return {
        ...state,
        current,
        cards: newCards
      };
    }

    // action new
    case 'new': {
      return {
        ...state,
        current: -1
      };
    }

    // action next
    case 'next': {
      // get cards and current index from state
      const {cards, current, show} = state;

      if(cards[current].subject === 'Math') {
        // do math stuff
        const nextMathCard: Card = getAdditionCard();
        cards.push(nextMathCard);
  
        return {
          ...state,
          cards,
          current: current+1
        };
      } else {
        // call to the getNext function
        const next = getNext({
          cards,
          current,
          show
        });

        return {
          ...state,
          current: next
        };
      }
    }

    // action save
    case 'save': {
      const {cards} = state;
      const {answer, question, subject} = action;

      // get the index of the card with this question
      // if there is no existing card with that question
      // index will be -1
      const index = cards.findIndex(card => card.question === question);

      // A card object with the values received in the action
      const card = {
        answer,
        question,
        subject
      } as Card;

      // create a new array of cards
      // filter out 'invalid' cards that don't have a question
      // aka filter for only cards that have a question
      const newCards = cards.filter(v => !!v.question);

      // if the question already exists in the array
      if(index > -1) {
        // assign the card object to the index
        newCards[index] = card;
      } else {
        // if the question does not already exist in the array
        // add the card object to the array
        newCards.push(card);
      }

      // return new context
      return {
        ...state,
        cards: newCards
      };
    }

    // action select
    case 'select': {
      const {cards} = state;
      const {question, subject} = action;

      const currentQuestion = question
        ? question
        : cards.filter((card) => card.subject === subject)[0].question;

      const current = cards.findIndex(card => card.question === currentQuestion);

      if(current < 0) return state;

      return {
        ...state,
        current
      };
    };

    // action showAdd
    case 'showAdd': {
      const {subject} = action;
      const show = [...state.show];

      !show.includes(subject) && show.push(subject);

      // add updated show to state
      Object.assign(state, {show});

      return {
        ...state,
        show
      };
    };

    // action showAll
    case 'showAll': 
      return {
        ...state,
        show: []
      };
    ;

    // action showRemove
    case 'showRemove': {
      const {subject} = action;
      const show = state.show.filter(subj => subj !== subject);

      return {
        ...state,
        show
      };
    };

    // default case returns the previous state without changing it
    default: 
      return state;
  }
}

// the object that we use to make the first Context
export const initialState = getInitialState();

// a context object made from initialState
const CardContext = createContext(initialState);

// the Props that the CardProvider will accept
type CardProviderProps = {
  // you can put react components inside of the Provider component 
  children: React.ReactNode;

  // We might want to pass a state into the CardProvider for testing purposes
  testState?: CardState
};

const CardProvider = ({children, testState}: CardProviderProps) => {
  // useReducer returns an array containing the state at [0]
  // and the dispatch method at [1]
  // use array destructuring to get state and dispatch
  const [state, dispatch] = useReducer(reducer, testState ? testState : initialState); 

  // hook to trigger a function anytime cards changes
  useEffect(() => {
    // save cards to localStorage
    saveCards(state.cards);
  }, [state.cards]);

  // value is an object created by spreading state
  // and adding the dispatch method
  const value = {...state, dispatch};

  return (
    // returns a Provider with the state and dispatch that we created above
    <CardContext.Provider value={value}>
      {children}
    </CardContext.Provider>
  );
}

export {
  // some components will import CardContext so they can access the state using useContext
  CardContext,
  // the App will import the CardProvider so the CardContext will be available to components
  CardProvider
};
