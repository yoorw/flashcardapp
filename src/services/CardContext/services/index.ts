import { getAdditionCard } from '../../../shared/utils';
import { Card, CardAction, CardState } from '../../../types';


// card objects
const card1: Card = {
  question: 'What is a linked list?',
  subject: 'Linked List',
  answer: `A linked list is a sequential list of nodes. 
  The nodes hold data.
  The nodes hold pointers that point ot other nodes containing data.`
};
const card2: Card = {
  question: 'What is a stack?',
  subject: 'Stack',
  answer: `A stack is a one ended linear data structure.
  The stack models real world situations by having two primary operations: push and pop.
  Push adds an element to the stack.
  Pop pulls the top element off the stack.`
};
const card3: Card = getAdditionCard();

export const cards = [card1, card2, card3];

// a function that loads the cards from localStorage
// and returns a CardState object
export const getInitialState = () => {

  return ({

    // the cards that are displayed to the user
    // if loadedCards is undefined, use cards
    // cards: loadedCards ? shuffle(loadedCards) : cards,       // ORIGINAL
    cards: cards,
  
    // index of the currently displayed card
    current: 0,
  
    // placeholder for the dispatch function
    dispatch: (action: CardAction) => undefined,
  
    // the array of subjects to show the user
    show: []
  } as CardState);
};

export const getNext = ({
  cards,
  current,
  show
}: {
  cards: Card[],
  current: number,
  show: string[]
}) => {
  // show array is empty, so we are showing all card
  if(show.length === 0) {
    const total = cards.length - 1;
    // just add 1, if +1 is too big return 0
    const next = current + 1 <= total
      ? current + 1
      : 0;

    return next;
  } else {
    // filter cards. Only keep cards with a subject that's in show
    const showCards = cards.filter((card) => show.includes(card.subject));

    // get the index of the current card in the showCards array
    const showCurrent = showCards.findIndex((card) => card.question === cards[current].question);

    const showTotal = showCards.length - 1;

    // showNext gives us the next index in the showCards array
    const showNext = showCurrent + 1 <= showTotal
      ? showCurrent + 1
      : 0;

    // translate the showNext index to the index of the same card in cards
    const next = cards.findIndex((card) => card.question === showCards[showNext].question);

    return next;
  }
};

export const getPrevious = ({
  cards,
  current,
  show
}: {
  cards: Card[],
  current: number,
  show: string[]
}) => {
  // show array is empty, so we are showing all cards
  if(show.length === 0) {
    // decrease current by 1, unless decrementing results in negative
    const previous = current - 1 >= 0
      ? current - 1
      : 0;

    return previous;
  } else {
    // filter cards. only keep cards with a subject that's in show
    const showCards = cards.filter((card) => show.includes(card.subject));

    // get the index of the current card in the showCards array
    const showCurrent = showCards.findIndex((card) => card.question === cards[current].question);

    // get previous index in showCards array
    const showPrevious = showCurrent - 1 >= 0
      ? showCurrent - 1
      : 0;

    // translate the showPrevious index to the index of the same card in cards
    const previous = cards.findIndex((card) => card.question === showCards[showPrevious].question);

    return previous;
  }
};
