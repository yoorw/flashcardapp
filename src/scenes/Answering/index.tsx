// import React to that we can use JSX
import React, {useContext, useEffect, useState} from 'react';
// import all the components from Semantic UI React
import {
  Button,
  Container,
  Form,
  Header,
  TextArea
} from 'semantic-ui-react';

import Answer from './components/Answer';
import Stats from './components/Stats';
import TrueStats from './components/TrueStats';
import { CardContext } from '../../services/CardContext';
import {StatsContext} from '../../services/StatsContext';
import { CardActionTypes, StatsActionType } from '../../types';
import { isCardFirstInSubject } from '../../shared/utils';


const Answering = () => {
  // get cards and current index from CardContext
  const {cards, current, dispatch, show} = useContext(CardContext);
  const {dispatch: statsDispatch} = useContext(StatsContext);

  
  // determine if Back button should be disabled
  const backDisabled: boolean = isCardFirstInSubject(cards, current, show);

  // TODO: !!! MOVE THIS FUNCTION - isCardFirstInSubject - to the disabled property of the back button below.
  // Then figure out how to get CardContext show property as an input to figure out if cards
  // from selected Subject should show or not 

  // get the question from the current card
  let {question} = cards[current];

  const [showAnswer, setShowAnswer] = useState(false);

  // the value of the textarea where the user types their input 
  const [input, setInput] = useState('');

  // compare if input is correct or not
  const inputResult: StatsActionType = input === cards[current].answer
    ? StatsActionType.right
    : StatsActionType.wrong;


  useEffect(() => {
    // hide the answer
    setShowAnswer(false);

    // clear the TextArea
    setInput('');

    // useEffect triggers when the value of current changes
  }, [current, setShowAnswer, setInput]);

  return (
    <Container data-testid='container' style={{position: 'absolute', left: 200}}>
      <Header data-testid='question'><Stats/>{question}</Header>
      <Button 
      disabled={backDisabled} 
      onClick={() => {
        dispatch({type: CardActionTypes.back});
      }}>Go Back</Button>
      <Button onClick={() => {
        dispatch({type: CardActionTypes.next});
        statsDispatch({type: StatsActionType.skip, question});
      }}>Next or Skip</Button>
      <Form>
        <TextArea data-testid='textarea'
        value={input}
        onChange={(e: any, {value}) => typeof(value) === 'string' && setInput(value)}/>
      </Form>
      <Button content='Submit' 
        onClick={() => {
          statsDispatch({type: inputResult, question})
          dispatch({type: CardActionTypes.next})
        }}/>
      <Answer visible={showAnswer}/>
      <TrueStats/>
    </Container>
  )};

  export default Answering;
