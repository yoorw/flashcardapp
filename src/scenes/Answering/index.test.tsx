// React lets us create and display components to the user 
// We need to import it so that we can look at the components to test them 
import React, {useContext} from 'react';

// testing library gives us methods to test components 
// we use render to look at React components
// we use cleanup to clear out memory after tests 
import { render, cleanup, fireEvent} from '@testing-library/react';

// extend-expect gives us methods that let us say what 
// we think a component will look like when we test it
import '@testing-library/jest-dom/extend-expect';

// This is the Answering component that we are going to write
// we have to import it so that we can look at it to test it
import Answering from './index';
import { CardState, StatsState } from '../../types';
import {CardProvider, initialState} from '../../services/CardContext';
import {StatsContext, StatsProvider} from '../../services/StatsContext';


afterEach(cleanup);

const renderAnswering = (testState?: CardState) => {
  return render(
    <CardProvider testState={testState ? testState : initialState}>
      <Answering />
    </CardProvider>
  );
};

// we need
// a container,
it('has a Container', () => {
  const {getByTestId} = render(<Answering/>);
  const container = getByTestId('container');
  expect(container).toBeInTheDocument();
});

// test to see if the question prompt is in the document
it('has a question prompt', () => {
  // Use Object Destructuring to get the getByTestId from the result of the render
  const {getByTestId} = render(<Answering/>);

  // find question by searching for testId 'question'
  const question = getByTestId('question');

  // assert that question is in the document
  expect(question).toBeInTheDocument();
});

// test to see if the question prompt is from the current card
it('has the question prompt from the current card', () => {
  const {cards, current} = initialState;

  // get the question from the current card
  const currentQuestion = cards[current].question;

  // get getByTestId from the helper function
  const {getByTestId} = renderAnswering();

  const question = getByTestId('question');

  // question content should be the question from the current card
  expect(question).toHaveTextContent(currentQuestion);
});

// Back Button
describe('Back button', () => {
  afterEach(cleanup);

  // test to see if the Back button is in the document
  it('has a button to go to the previous card', () => {
    // Use Object Destructuring to get getByText from the result of the render
    const {getByText} = render(<Answering/>);

    // find Back button by search for string 'Go Back'
    const back = getByText('Go Back');

    // assert Back button is in the document
    expect(back).toBeInTheDocument();
  });

  it('Back button should be disabled when current === 0', () => {
    // create a new CardState with current === 0
    const zeroState = {
      ...initialState,
      current: 0
    };

    const {getByTestId, getByText} = renderAnswering(zeroState);

    // get backButton by text - users find buttons with text
    const backButton = getByText(/back/i);
    
    expect(backButton.closest('button')).toBeDisabled();
  });
});

// test to see if the Skip button is in the document
it('has a button to skip the card', () => {
  // Use Object Destructuring to get getByText from the result of the render
  const {getByText} = render(<Answering/>);

  // find Skip button by searching for string 'Skip'
  const skip = getByText('Next or Skip');

  // assert that Skip button is in the document
  expect(skip).toBeInTheDocument();
});

// a textarea to write the answer in
it('has a textArea to type the answer in', () => {
  const {getByTestId} = render(<Answering/>);
  const textArea = getByTestId('textarea');
  expect(textArea).toBeInTheDocument();
});

// test to see if the Submit button is in the document
it('has a button to submit the answer', () => {
  const {getByText} = render(<Answering/>);
  const submit = getByText('Submit');
  expect(submit).toBeInTheDocument();
});

// test that the skip button works
it('clicks the skip button and the next question appears', () => {
  // create a CardState with current set to 0
  const zeroState = {
    ...initialState,
    current: 0
  };

  // current starts at 0
  const {getByTestId, getByText} = renderAnswering(zeroState);

  const question = getByTestId('question');
  // current starts out at 0, so question should be cards[0]
  expect(question).toHaveTextContent(initialState.cards[0].question);

  const skip = getByText(/next or skip/i);
  // this should change current index from 0 to 1
  fireEvent.click(skip);

  expect(question).toHaveTextContent(initialState.cards[1].question);
});

it('clears the answer when card changes', () => {
  const {getByText, getByTestId} = renderAnswering();
  const textarea = getByTestId('textarea');

  const placeholder = 'placeholder text';
  // put the placeholder text into the textarea
  fireEvent.change(textarea, {target: {value: placeholder}});

  // make sure the placeholder text is there
  expect(textarea).toHaveTextContent(placeholder);

  // get the skip button
  const skip = getByText(/next or skip/i);
  // click skip, this dispatches a 'next' action to cardcontext
  // which should change the value of current
  // and trigger useEffect hook to clear the textarea
  fireEvent.click(skip);

  // textarea should be empty
  expect(textarea).toHaveTextContent('');
});

it('has the stats icon', () => {
  const {getByTestId} = renderAnswering();
  const stats = getByTestId('icon');
  expect(stats).toBeInTheDocument();
});

// when the user clicks the skip button, the skip is recorded in the stats
it('clicking skip records stats', () => {
  // create a CardState with current set to 0
  const cardState = {
    ...initialState,
    current: 0
  };

  // a blank stats object
  const blankStats = {
    right: 0,
    wrong: 0,
    skip: 0
  };

  // get the question from cards index 0
  const {question} = cardState.cards[0];

  // create statsState with stats for the question
  const statsState: StatsState = {
    [question]: blankStats
  };

  // helper component displays the value of skip for the question
  const SkipDisplay = () => {
    const stats = useContext(StatsContext);
    const {skip} = stats[question];
    return <div data-testid='skipDisplay'>{skip}</div>;
  };

  // render Answering and SkipDisplay inside the providers
  // pass the providers the cardState and StatsState values that we defined
  const {getByTestId, getByText} = render(
    <CardProvider testState={cardState}>
      <StatsProvider testState={statsState}>
        <Answering />
        <SkipDisplay />
        </StatsProvider>
    </CardProvider>
  );

  // find the skip button
  const skipButton = getByText(/next or skip/i);

  // find the skip display
  const skipDisplay = getByTestId('skipDisplay');

  // skip display should start at 0
  expect(skipDisplay).toHaveTextContent('0');

  // click the skip button
  fireEvent.click(skipButton);

  expect(skipDisplay).toHaveTextContent('1');
});

// and the snapshot 
it('Matches Snapshot', () => {
  // get the asFragment method so we can look at the component as a DocumentFragment
  const {asFragment} = render(<Answering/>);

  // expect the result of asFragment() to match the snapshot of this component
  expect(asFragment()).toMatchSnapshot();
});

describe('submit button controls display of the answer', () => {
  // the answer to the current question
  const initialAnswer = initialState.cards[initialState.current].answer;
  const withoutLineBreaks = initialAnswer.replace(/\s{2,}/g, " ");
  const compareToInitialAnswer = (
    content: string,
    {textContent}: HTMLElement
  ) => !!textContent &&
    textContent
    .replace(/\s{2,}/g, " ")
    .slice(6, textContent.length) === withoutLineBreaks;

  // answer does not show up
  it('the answer does not show up before the submit button is clicked', () => {
    const {queryByText} = renderAnswering();

    // use the custom function to search for the initial answer
    const answer = queryByText(compareToInitialAnswer);

    expect(answer).toBeNull();
  });

  // clicking submit button skips to the next question 
  it('clicks the skip button and the next question appears', () => {
    // create a CardState with current set to 0
    const zeroState = {
      ...initialState,
      current: 0
    };

    // current starts at 0
    const {getByTestId, getByText} = renderAnswering(zeroState);

    const question = getByTestId('question');
    // current starts out at 0, so question should be cards[0]
    expect(question).toHaveTextContent(initialState.cards[0].question);

    const submit = getByText(/submit/i);
    // this should change current index from 0 to 1
    fireEvent.click(submit);

    expect(question).toHaveTextContent(initialState.cards[1].question);
  });
});
