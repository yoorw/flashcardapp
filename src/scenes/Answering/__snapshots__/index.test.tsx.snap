// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches Snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui container"
    data-testid="container"
    style="position: absolute; left: 200px;"
  >
    <div
      class="ui header"
      data-testid="question"
    >
      <i
        aria-hidden="true"
        class="question circle icon"
        data-testid="icon"
      />
      What is a linked list?
    </div>
    <button
      class="ui disabled button"
      disabled=""
      tabindex="-1"
    >
      Go Back
    </button>
    <button
      class="ui button"
    >
      Next or Skip
    </button>
    <form
      class="ui form"
    >
      <textarea
        data-testid="textarea"
        rows="3"
      />
    </form>
    <button
      class="ui button"
    >
      Submit
    </button>
    <h3
      class="ui header"
      data-testid="truestats"
    >
      <div>
        <div>
           Total Questions: 0
        </div>
        <div>
           Questions Right: 0
        </div>
        <div>
           Questions Wrong: 0
        </div>
        <div>
           Questions Skipped: 0
        </div>
      </div>
    </h3>
  </div>
</DocumentFragment>
`;
